auction-server/src/auction.rs

```diff
@@ -1066,3 +1034,353 @@ async fn simulate_bid_svm(chain_store: &ChainStoreSvm, bid: &BidSvm) -> Result<(
+                    result: _,
+                    reason: _,
+                }) => {}
+                // Either simulation was successful or we can't simulate at this moment
+                _ => return Ok(vec![bid.clone()]),
+            }
+        }
+        Ok(vec![])
+    }
+
+    #[tracing::instrument(skip_all)]
+    async fn submit_bids(
+        &self,
+        _permission_key: Bytes,
+        bids: Vec<Self::SimulatedBid>,
+        store: Arc<Store>,
+    ) -> Result<Vec<u8>> {
+        let relayer = store.express_relay_svm.relayer.clone();
+        let mut bid = bids[0].clone();
+        let serialized_message = bid.transaction.message.serialize();
+        let relayer_signature_pos = bid
+            .transaction
+            .message
+            .static_account_keys()
+            .iter()
+            .position(|p| p.eq(&relayer.pubkey()))
+            .expect("Relayer not found in static account keys");
+        bid.transaction.signatures[relayer_signature_pos] =
+            relayer.sign_message(&serialized_message);
+        match self.client.send_transaction(&bid.transaction).await {
+            Ok(response) => Ok(response.as_ref().to_vec()),
+            Err(e) => {
+                tracing::error!("Error while submitting bid: {:?}", e);
+                Err(anyhow!(e))
+            }
+        }
+    }
+
+    async fn get_bid_results(
+        &self,
+        bids: Vec<Self::SimulatedBid>,
+        tx_hash: Vec<u8>,
+    ) -> Result<Option<Vec<BidStatus>>> {
+        if bids.len() != 1 {
+            return Err(anyhow!("Invalid number of bids: {}", bids.len()));
+        }
+
+        let status = self
+            .client
+            .get_signature_status_with_commitment(
+                &SignatureSvm::try_from(tx_hash.clone())
+                    .map_err(|e| anyhow!("Invalid svm signature: {:?}", e))?,
+                CommitmentConfig::confirmed(),
+            )
+            .await?;
+
+        match status {
+            Some(res) => Ok(Some(vec![match res {
+                Ok(()) => BidStatus::Won {
+                    index:  0,
+                    result: tx_hash,
+                },
+                Err(_) => BidStatus::Lost {
+                    index:  Some(0),
+                    result: Some(tx_hash),
+                },
+            }])),
+            None => {
+                // not yet confirmed
+                Ok(None)
+            }
+        }
+    }
+}
+
+async fn run_submission_loop<T: Auction>(
+    store: Arc<Store>,
+    chain_store: T,
+    chain_id: String,
+) -> Result<()> {
+    tracing::info!(chain_id = chain_id, "Starting transaction submitter...");
+    let mut exit_check_interval = tokio::time::interval(EXIT_CHECK_INTERVAL);
+
+    let ws_client = chain_store.get_ws_client().await?;
+    let mut stream = T::get_block_stream(&ws_client).await?;
+
+    while !SHOULD_EXIT.load(Ordering::Acquire) {
+        tokio::select! {
+            block = stream.next() => {
```

Comment:
this is using next block with transactions as a signal that a slot has passed, but slots can be skipped or have empty blocks. it may be better to get pings when a slot has passed (even when there is no block), or to just directly query the slot and update the last slot number

Reply:
We just want a new trigger to run auctions. It's not importanat what create this trigger and while the conclusion and submission being triggered for every block, we are good. Actually we don't use the block value in the process.

Reply:
yes but this trigger is that the block is Some? maybe make sure that this stream will actually return Some even if the block doesn't get created, e.g. bc the slot is skipped

Reply:
It seems that it's using this rpc method: https://solana.com/docs/rpc/http/getblock
So If the block is confimed, it can't be null.

What do you think?

Reply:
Based on this link, the notification is going to be triggered when a block is confirmed or finalized.
https://solana.com/docs/rpc/websocket/blocksubscribe